### PR TITLE
PR: Use JupyterLab to render notebooks

### LINKED ISSUE #261: Transition to JupyterLab
The old Jupyter notebook is no longer maintained, so we should try to use JupyterLab instead.

The notebook interface is at https://github.com/jupyterlab/jupyterlab/tree/master/packages/notebook . There is an example demonstrating a stand-alone "notebook" at https://github.com/jupyterlab/jupyterlab/tree/master/examples/notebook .

### PR DESCRIPTION
**Pull request by jitseniesen**
<s>This is far from ready; I just want to show my progress.</s>

This PR updates the plugin to use the JupyterLab framework instead of the classic notebook interface, because JupyterLab is the future.

To do list:
- [x] Allow for different notebook names instead of assuming it's called `test.ipynb`
- [x] Add the buttons for Save, Insert Cell, Cut, Copy, etc.
- [x] Fix saving of notebook when user closes it
- [x] Don't ask user to select kernel when opening a new notebook
- [x] <s>Remove sidebar? Add menubar?</s> deferred to issue #270 
- [x] Make sure unit tests pass
- [x] Update CI script
- [x] Write contribution info

Fixes #261

### CODE REVIEW

**Comment by jitseniesen** [comment]
This is just a proof of concept to show off a screen shot:
Screenshot_20200327_160014
This makes me more confident that I can make it work!

**Comment by goanpeca** [comment]
Amazing work @jitseniesen !!! Keep it up :-)

**Comment by ccordoba12** [comment]
Pretty nice!! I just have one question: is this going to create one server instance per notebook?

**Comment by jitseniesen** [comment]
> Pretty nice!! I just have one question: is this going to create one server instance per notebook?

I am planning to use the same approach as currently at the start, which is one server instance per directory. However, I want to look at whether it is possible to have only one server instance which can render all notebooks regardless of where they are stored. Unless you think it's preferable to have one server instance per notebook.

**Comment by ccordoba12** [comment]
> I am planning to use the same approach as currently at the start, which is one server instance per directory

The current approach is really one server per user home directory. So if you want to open a notebook in `/tmp`, you need a new server. I decided to do that to not expose the entire filesystem to the server.

> However, I want to look at whether it is possible to have only one server instance which can render all notebooks regardless of where they are stored

That probably won't work on Windows, in case users want to load notebooks from `C:\`, `D:\`, etc.

**Comment by jitseniesen** [comment]
Ok, I won't change that in this PR, let's try to stick to one server per home dir.

**Comment by jitseniesen** [comment]
Most things seem to work:

spyder-notebook-2020-03-31_13 58 58

**Comment by ccordoba12** [comment]
> Most things seem to work:

That looks pretty neat!! Thanks @jitseniesen for working on this!

But what about the connection with the IPython console? Is that working?

**Comment by jitseniesen** [comment]
> But what about the connection with the IPython console? Is that working?

I tried it out and the old code still seems to work, which is a big relief. You can open a console, access the variables from the notebook in the console and in the variable explorer, and changes made in the console are reflected in the notebook. Let me know if I need to test anything else.

**Comment by ccordoba12** [comment]
@jitseniesen, you can remove support for Python 2 in version 0.3, if you want.

File: spyder_notebook/widgets/client.py [hunk 1]
```diff
@@ -214,8 +205,19 @@ def get_short_name(self):
...
+        """
+        Save current notebook.
+
+        This function saves the current notebook by simulating a click on the
+        Save button in the notebook. The Save button is found by selecting the
+        first element of class `jp-ToolbarButtonComponent` whose `title`
+        attribute begins with the string "Save".
+
+        NB: I am not sure whether the save is performed before the function
+        returns.
```
Comment (ccordoba12) [inline spyder_notebook/widgets/client.py:217]: This is probably not true because almost all JS operations are asynchronous.

**Comment by jitseniesen** [comment]
I want to get rid of the side bar on the left with the commands and replace it with a menu bar, accessing the same commands (and more). However, I am thinking to leave that to another PR.

Apart from that, I need to write some documentation while things are still fairly fresh in my mind, and then I think this is ready.

**Comment by ccordoba12** [comment]
> I want to get rid of the side bar on the left with the commands and replace it with a menu bar, accessing the same commands (and more). However, I am thinking to leave that to another PR.

I'm fine with that.

For a future PR you should also consider to add some extra JupyterLab extensions. For instance, the ones necessary to display Bokeh or Plotly plots.

**Pull request merged**